---
title: "Cox analysis"
output: html_document
date: "2025-02-05"
---

This script calculates individual cox Hazard ratios on each drought-adapted sites. 


```{r setup, include=FALSE}
rm(list= ls())

library(survival)
library(tidyr)
library(dplyr)
library(car)

setwd("/Users/rozenn/Library/CloudStorage/GoogleDrive-rozennpineau@uchicago.edu/My Drive/Work/9.Science/1.DroughtProject/1.analyses/data/6.trajectories/clumped")
phenos <- read.table("phenos_env.txt", header = T)

#reshape file to have genotypes as columns with 1/0 
phenos_wide <- phenos %>%
  mutate(value = 1) %>%  # Add a column with value 1 to denote presence
  pivot_wider(names_from = pair, values_from = value, values_fill = 0)

#add column that specifies the event type (death)
phenos_wide$event <- 1 #everyone is dead at day to full wilt (1 for event/death, 0 for censored)

# View the reshaped data
print(phenos_wide)

```

The summary output includes the coefficients (`coef`), their exponentiated values (hazard ratios, `exp(coef)`), standard errors (`se(coef)`), Z-values (`z`), and p-values (`Pr(>|z|)`). Here's how to interpret these:

1. **Coefficient (`coef`):** The coefficient represents the log hazard ratio. A positive coefficient implies that with an increase in the predictor variable, the hazard (risk of event) increases.
2. **Exponentiated Coefficient (`exp(coef)`):** The hazard ratio indicates the change in hazard for a one-unit increase in the predictor variable. Values greater than 1 indicate increased risk, whereas values less than 1 indicate reduced risk.


```{r adding_ancestry}
#upload dataset
setwd("/Users/rozenn/Library/CloudStorage/GoogleDrive-rozennpineau@uchicago.edu/My Drive/Work/9.Science/1.DroughtProject/1.analyses/data/8.ancestry_hmm/1.results/trajectories/")
genotype <- read.table("two_pulse_flexible_prop_2_values_ID_filtered_GT.txt")
samp_order <- t(read.table("samp_order.txt")) #sample order from the GT file

#upload admixture data
adm_full <- read.table("/Users/rozenn/Library/CloudStorage/GoogleDrive-rozennpineau@uchicago.edu/My Drive/Work/9.Science/1.DroughtProject/1.analyses/data/2.admixture/k2_structure_res.csv", sep = ",", header = T)
adm <- data.frame(full_ID = adm_full$samp, admx = adm_full$k2.V1) #proportion of one genotype


#remove outliers
idx <- c( which(samp_order == "P16_Nat_1_T") , which(samp_order == "P12_Nat_14_T") ) 
samp_order_fil <- data.frame(samp = samp_order[-idx])
gt <- genotype[,-c(1:4)]
gt <- gt[,-c(idx)] #35x280
#in adm
adm_fil <- adm[adm$full_ID !="P16_Nat_1_T" & adm$full_ID != "P12_Nat_14_T",] #280 x  2
```


```{r make_ancestry_matrix}
#metrics
n_samp <- dim(gt)[2]
n_snp <- dim(gt)[1]

#make ancestry matrix
snp <- matrix(NA, n_snp, n_samp) #35 x 280

for (x in 1:n_snp) {
  
  for (y in 1:n_samp) {
    
    if (gt[x,y] == "0|0") {snp[x,y] <- 0} #homs reference
    if (gt[x,y] == "0|1" | gt[x,y] == "1|0" ) {snp[x,y] <- 1} #hets
    if (gt[x,y] == "1|1") {snp[x,y] <- 2} #homs alternative
    if (gt[x,y] == "./.") {snp[x,y] <- NA} #NA
    
  }
}

#prep data for Cox survival object
cox_anc_dt <- data.frame(cbind(samp_order_fil, t(snp)))
colnames(cox_anc_dt) <- c("full_ID", c(paste("snp", 1:35, sep="")))
phenos_wide_adm <- merge(phenos_wide, adm_fil, by = "full_ID", sort = F) #merge phenos_wide with admixture proportions
cox_anc_dt_full <- merge(cox_anc_dt, phenos_wide_adm, by = "full_ID", sort = F) #merge with ancestry


#clean
rm(cox_anc_dt, adm, adm_full, phenos_wide_adm, gt, genotype)
```

I tried a model on all snps at the same time but it did not converge. I thus test each SNP individually and adjust p-value with FDR correction.

```{r cox_on_ancestry}
#create the survival object
survival_obj <- Surv(time = cox_anc_dt_full$survival, event = cox_anc_dt_full$event)

# Define the Cox model
cox_model <- coxph(survival_obj ~ snp1 +  snp2 +  snp3 +  snp4 +  snp5 +  snp6 +  snp7 +  snp8 +  snp9 +  snp10 + snp11 + snp12 + snp13 + snp14 + snp15 + snp16 + snp17 + snp18 + snp19 + snp20 + snp21 + snp22 + snp23 + snp24 + snp25 + snp26 + snp27 + snp28 + snp29 + snp30 + snp31 + snp32 + snp33 + snp34 + snp35 + admx, data = cox_anc_dt_full, control = coxph.control(iter.max = 1000))
# I am having an error :  Loglik converged before variable  1,2,3,4,6,8,10,12,13,14,15,16,18,22,30 ; coefficient may be infinite.# the model does not converge with all snps together

#reduce the model to test each site with genome wide ancestry (admixture)
store <- data.frame(site = 1:n_snp, exp_coef = rep(NA, n_snp), se_coef = rep(NA, n_snp), p_val = rep(NA, n_snp), 
                    exp_coef_admx = rep(NA, n_snp), se_coef_admx = rep(NA, n_snp), p_val_admx = rep(NA, n_snp))

for (site in 1:n_snp) {
  
  cox_model <- coxph(survival_obj ~ cox_anc_dt_full[,1+site] + admx, data = cox_anc_dt_full)
  #fill data table
  store$exp_coef[site] <- summary(cox_model)$coefficients[1,2] #exponentiated coef for site
  store$se_coef[site] <- summary(cox_model)$coefficients[1,3] #standard error for exp coef for site
  store$p_val[site] <- summary(cox_model)$coefficients[1,5] #pval for site
  store$exp_coef_admx[site] <- summary(cox_model)$coefficients[2,2] #exponentiated coef for covariate
  store$se_coef_admx[site] <- summary(cox_model)$coefficients[2,3] #standard error for exp coef for covariate
  store$p_val_admx[site] <- summary(cox_model)$coefficients[2,5] #pval for covariate
  
  
}

# Exponentiated Coefficient (exp(coef)): The hazard ratio indicates the change in hazard for a one-unit increase in the predictor variable. Values greater than 1 indicate increased risk, whereas values less than 1 indicate reduced risk.

```

```{r visualize}

par(family = "Times New Roman")
hist(store$exp_coef[store$p_val < 0.05], xlab = "Exponentiated Cox coefficient (risk)", main = "")

length(which(store$fdr_p_val < 0.05)) # 14 significant sites

#after False Discovery Rate correction
store$p_val_fdr <- p.adjust(store$p_val, method = "fdr")

pdf(file=paste("/Users/rozenn/Library/CloudStorage/GoogleDrive-rozennpineau@uchicago.edu/My Drive/Work/9.Science/1.DroughtProject/0.Writing/1.Paper/1.Figures/Figure3/cox.pdf", sep="" ), bg = "transparent", width=3, height=4, family = "Times New Roman")


length(which(store$p_val_fdr < 0.05)) # 14 significant sites
par(family = "Times New Roman", cex.axis = 1.5, cex.lab = 1.5)
hist(store$exp_coef[store$p_val_fdr < 0.05], 14, xlab = "Exponentiated Cox coefficient (risk)", main = "")

```
```{r visualize_v2}

#after False Discovery Rate correction
store$p_val_fdr <- p.adjust(store$p_val, method = "fdr")
which(store$p_val_fdr < 0.05)

length(which(store$p_val_fdr < 0.05)) # 14 significant sites
length(which(store$p_val < 0.05)) # 20 significant sites

idx <- store$p_val_fdr < 0.05


pdf(file="/Users/rozenn/Library/CloudStorage/GoogleDrive-rozennpineau@uchicago.edu/My Drive/Work/9.Science/1.DroughtProject/0.Writing/1.Paper/1.Figures/Figure3/cox_all.pdf", 
    bg = "transparent", width=4.5, height=3, family = "Times New Roman")

par(family = "Times New Roman", cex.lab = 1.5, cex.axis = 1.5)
hist(store$exp_coef[!idx], breaks = 20, xlab = "Hazard Ratio", main = "", xlim = c(0.5,.95))
hist(store$exp_coef[idx], breaks =15,add = T, col = "red")
box()

dev.off()

summary(store$exp_coef)


pdf(file=paste("/Users/rozenn/Library/CloudStorage/GoogleDrive-rozennpineau@uchicago.edu/My Drive/Work/9.Science/1.DroughtProject/0.Writing/1.Paper/1.Figures/Figure3/cox.pdf", sep="" ), bg = "transparent", width=3, height=4, family = "Times New Roman")


length(which(store$p_val_fdr < 0.05)) # 14 significant sites
par(family = "Times New Roman", cex.axis = 1.5, cex.lab = 1.5)
hist(store$exp_coef[store$p_val_fdr < 0.05], 14, xlab = "Exponentiated Cox coefficient (risk)", main = "")

```
